src/clj/rems/ega.clj

```diff
@@ -0,0 +1,226 @@
+(ns rems.ega
+  "Utilities for interfacing with EGA Permission API.
+
+  The function names match the remote API paths."
+  (:require [clj-http.client :as http]
+            [clj-time.core :as time-core]
+            [clojure.string :as str]
+            [clojure.test :refer [deftest is]]
+            [clojure.tools.logging :as log]
+            [rems.config :refer [env]]
+            [rems.db.user-secrets :as user-secrets]
+            [rems.ga4gh :as ga4gh]
+            [rems.json :as json]))
+
+(defn token
+  "Fetches an EGA Token representing an EGA user.
+
+  `:username`      - username (in EGA) e.g. foo@bar.com
+  `:password`      - password for the user
+  `:client-id`     - client id for REMS
+  `:client-secret` - client secret for REMS
+  `:config`        - configuration of the EGA integration
+
+  NB: It is valid for one hour."
+  [{:keys [username password client-id client-secret config]}]
+  (-> (http/post (str (:connect-server-url config) "/token")
+                 {:form-params {"grant_type" "password"
+                                "client_id" client-id
+                                "client_secret" client-secret
+                                "username" username
+                                "password" password
+                                "scope" "openid"}
+                  :socket-timeout 2500
+                  :conn-timeout 2500})
+      :body
+      json/parse-string))
+
+(defn api-key-generate
+  "Generate an API-Key for a user with a token.
+
+  The API-Key can be used longer (e.g. a year) to represent the
+  user it was generated for.
+
+  `:access-token`    - the access token
+  `:id`              - identity for the key e.g. user
+  `:expiration-date` - YYYY-MM-DD of the desired expiration date
+  `:reason`          - string description of the use of the key
+  `:config`          - configuration of the EGA integration"
+  [{:keys [access-token id expiration-date reason config]}]
+  (-> (http/get (str (:permission-server-url config) "/api_key/generate")
+                {:oauth-token access-token
+                 :query-params {"id" id
+                                "expiration_date" expiration-date ; TODO: pass as date?
+                                "reason" reason}
+                 :socket-timeout 2500
+                 :conn-timeout 2500})
+      :body
+      json/parse-string))
+
+(defn api-key-list
+  "List the API-Keys available.
+
+  `:access-token` - the access token
+  `:config`       - configuration of the EGA integration
+
+  NB: The token itself cannot be returned anymore."
+  [{:keys [access-token config]}]
+  (-> (http/get (str (:permission-server-url config) "/api_key")
+                {:oauth-token access-token
+                 :socket-timeout 2500
+                 :conn-timeout 2500})
+      :body
+      json/parse-string))
+
+(defn read-permissions
```

Comment:
Inconsistency: This and the next function are prefixed `read-` even though it's not in the api path. Other GET requests aren't prefixed `read-`.

Reply:
I like naming the functions according to the API endpoints, and I think some sort of get- / read- / fetch- prefix might work here, but it needs to be applied consistently.

Reply:
You are right, I think I will prefix all the methods with the request-method so that they become unique. Otherwise GET permissions and DELETE permissions would be the same.

Reply:
They can also all be changed into private?